sys/linux/dev_dri.txt

```diff
@@ -453,13 +450,31 @@ drm_irq_busid {
 	func	int32
 }
 
-drm_map {
-	off	vma[opt]
-	size	intptr
-	type	flags[drm_map_type, int32]
-	flags	flags[drm_map_flags, int32]
-	handle	vma
-	mtrr	int32
+drm_map$DRM_IOCTL_GET_MAP {
+	off	vma[opt]	(inout)
+	size	const[0, intptr]	(out)
+	type	const[0, int32]	(out)
+	flags	const[0, int32]	(out)
+	handle	vma	(out)
+	mtrr	int32	(out)
+}
+
+drm_map$DRM_IOCTL_ADD_MAP {
+	off	vma	(in)
+	size	intptr	(in)
+	type	flags[drm_map_type, int32]	(in)
+	flags	flags[drm_map_flags, int32]	(in)
+	handle	vma	(out)
+	mtrr	int32	(out)
+}
+
+drm_map$DRM_IOCTL_RM_MAP {
+	off	const[0, intptr]
+	size	const[0, intptr]
+	type	const[0, int32]
+	flags	const[0, int32]
+	handle	vma	(in)
```

Comment:
This structure is already passed as `in`:
```
ioctl$DRM_IOCTL_RM_MAP(fd fd_dri, cmd const[DRM_IOCTL_RM_MAP], arg ptr[in, drm_map$DRM_IOCTL_RM_MAP])
```
so it seems there is something wrong here.

Reply:
Only the `handle` field is used, which is `in`.

I just wanted to use per-field direction here to indicate this and to keep it consistent with the other options for `drm_map`. Removing it shouldn't make any difference since as you said the struct itself is passed as `in` already.

To avoid confusions, maybe it is best to remove the per-field directions for places where the upper-level ptr direction is enough to cover overall direction.

Another question (for here and other places): if some field is unused (e.g., `flags` is not used in this `drm_map`, it is neither checked so any value is valid), it is preferred to make it `const[0, T]` to avoid having fuzzer mutate values for this, right?

Reply:
> I just wanted to use per-field direction here to indicate this and to keep it consistent with the other options for drm_map.

I see.

> To avoid confusions, maybe it is best to remove the per-field directions for places where the upper-level ptr direction is enough to cover overall direction.

Hard to say.
Using ptr direction for the whole struct when it's all in or out is what have in 99% of cases, so I assumed that an explicit direction is used to override the ptr direction.

Reply:
> it is preferred to make it const[0, T]

Yes.

Reply:
Now the per-field direction is removed and the ptr direction is used as before (and as how it is used all over the descriptions). ([here](https://github.com/google/syzkaller/pull/2008/files#diff-26102468373666679c2ae06107ae0d6bR478))